### PR TITLE
rand: implement an unbiased random integer from a range

### PR DESCRIPTION
**Pull request by paulidale**
Refer: https://github.com/apple/swift/pull/39143 for a description of the algorithm.

It is optimal in the sense of having:

* no divisions
* minimal number of blocks of random bits from the generator

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by t8m — COMMENTED** [review]
Should we also have a convenience function with lower bound? I.e., `lower + ossl_rand_uniform_uint32(upper - lower)` ? Or just add lower parameter here?

**Comment by paulidale** [comment]
[0, max) is the commonest range required.

A helper function is a reasonable idea still.

**Comment by sftcd** [comment]
I temporarily copied this into my [HPKE code](https://github.com/openssl/openssl/pull/22493/commits/83853add3aa3e577f76f2e9e36d44e984f723916) and it seems to work when picking GREASE values.

There's a snippet of code I used to informally test this [here](http://down.dsg.cs.tcd.ie/test_oruu.c). Not sure how one could write a good test that could be part of the test harness without generating some bogus test fails.

**Comment by paulidale** [comment]
The pathological case will be around 2³¹+1 where just under half the values are in the rejection region.  Given the test only needs the third random sample less than one time in 2³², testing isn't practical using less than (at a guess) 2⁶⁴ samples.

It would be possible to test this better by reducing the size.  E.g. generating 4 bit numbers under 9.  A mistake should become obvious in reasonable time.  Not useful for our test suite but good for confirmation.  χ² is probably good enough here.

**Comment by paulidale** [comment]
I've run a χ² test (p=0.99) for the output of this algorithm modified to work with 4 bit "words" and it seems good.  Test harness is [attached](https://github.com/openssl/openssl/files/13172348/test_oruu.c.txt).

There seems to be a slight problem with the χ² test for this discrete uniform distribution.  I'm seeing test failures for small samples at a higher than expected rate (gut feel, not measured).  For large samples it comes good and the test statistic reduces nicely.  I'm pretty sure this is because the sum of uniform distributions is pretty non-normal in nature.  As the number increases the central limit theorem kicks in and it tends towards normal.  Convergence to normal being, aproportional to the square root of the number of samples accumulated (i.e. 100 gives approx 1 digit, 10,000 gives approx 2 digits, 1,000,000 gives approx 3, etc).

Given the instability for small sample sizes and the time required to generate large samples, I do not see how this can reasonably be directly tested as part of our CIs.

Run using 1,000,000 samples (note the `jmax` of `-1` for powers of two which means no extra sampling was required which is optimal):
```
max =   2: 5000209,4999791,
max: 2, chi_sq: 0.017472, exp: 5000000.000000, jmax: -1
max =   3: 3334176,3333923,3331901,
max: 3, chi_sq: 0.932812, exp: 3333333.333333, jmax: 5
max =   4: 2499708,2501113,2499781,2499398,
max: 4, chi_sq: 0.693759, exp: 2500000.000000, jmax: -1
max =   5: 1999827,1998874,2000430,2001946,1998923,
max: 5, chi_sq: 3.214775, exp: 2000000.000000, jmax: 5
max =   6: 1668577,1665917,1664706,1666297,1667846,1666657,
max: 6, chi_sq: 5.749897, exp: 1666666.666667, jmax: 5
max =   7: 1430465,1428065,1427089,1431038,1428160,1427327,1427856,
max: 7, chi_sq: 10.047356, exp: 1428571.428571, jmax: 6
max =   8: 1249737,1249194,1249188,1250900,1250879,1247944,1250256,1251902,
max: 8, chi_sq: 8.696853, exp: 1250000.000000, jmax: -1
max =   9: 1110782,1112213,1110730,1109803,1111333,1110297,1112125,1110449,1112268,
max: 9, chi_sq: 6.026075, exp: 1111111.111111, jmax: 6
max =  10: 1001779,999831,1000708,999221,999167,999181,1000838,1000199,999904,999172,
max: 10, chi_sq: 7.102802, exp: 1000000.000000, jmax: 6
max =  11: 909015,909267,908319,909278,909771,909127,909099,909859,908548,909467,908250,
max: 11, chi_sq: 3.151274, exp: 909090.909091, jmax: 5
max =  12: 833981,833851,832271,833759,831721,834955,833605,831482,833055,833270,833272,834778,
max: 12, chi_sq: 15.480190, exp: 833333.333333, jmax: 5
max =  13: 769947,769383,769361,768888,769172,768927,770154,770032,769173,768483,769544,769447,767489,
max: 13, chi_sq: 7.802330, exp: 769230.769231, jmax: 6
max =  14: 714751,714341,715299,715759,715366,714639,715462,714295,713665,713116,713868,713069,712071,714299,
max: 14, chi_sq: 20.168345, exp: 714285.714286, jmax: 5
max =  15: 666945,666580,666357,666583,666835,665888,666941,666078,666901,667459,666406,667807,666785,665420,667015,
max: 15, chi_sq: 7.477271, exp: 666666.666667, jmax: 5
```

Run using 100,000,000 samples:
```
max =   2: 50000227,49999773,
max: 2, chi_sq: 0.002061, exp: 50000000.000000, jmax: -1
max =   3: 33333736,33335688,33330576,
max: 3, chi_sq: 0.399284, exp: 33333333.333333, jmax: 6
max =   4: 24999959,24991483,25010198,24998360,
max: 4, chi_sq: 7.169191, exp: 25000000.000000, jmax: -1
max =   5: 19997333,19999866,19998849,20005677,19998275,
max: 5, chi_sq: 2.182980, exp: 20000000.000000, jmax: 6
max =   6: 16662808,16667571,16670404,16665877,16662650,16670690,
max: 6, chi_sq: 3.757151, exp: 16666666.666667, jmax: 6
max =   7: 14292468,14286611,14284433,14281405,14284730,14290095,14280258,
max: 7, chi_sq: 8.159124, exp: 14285714.285714, jmax: 7
max =   8: 12501541,12502211,12498975,12494419,12499109,12498508,12503983,12501254,
max: 8, chi_sq: 4.793451, exp: 12500000.000000, jmax: -1
max =   9: 11115182,11112197,11108881,11111277,11111626,11110032,11107010,11109843,11113952,
max: 9, chi_sq: 4.561170, exp: 11111111.111111, jmax: 6
max =  10: 10003841,9996950,9997881,10000856,10006798,9996979,9997489,10003917,9998301,9996988,
max: 10, chi_sq: 11.822468, exp: 10000000.000000, jmax: 6
max =  11: 9082970,9093585,9093643,9088370,9092530,9093158,9090855,9093816,9090446,9092860,9087767,
max: 11, chi_sq: 12.555633, exp: 9090909.090909, jmax: 7
max =  12: 8333088,8336416,8331807,8332156,8332507,8333341,8331224,8337499,8332227,8336944,8327919,8334872,
max: 12, chi_sq: 9.804860, exp: 8333333.333333, jmax: 6
max =  13: 7696332,7692870,7692381,7688343,7697700,7691600,7689949,7693650,7697364,7690380,7694020,7684898,7690513,
max: 13, chi_sq: 20.737228, exp: 7692307.692308, jmax: 7
max =  14: 7145699,7144015,7140442,7139170,7146915,7145519,7140261,7147240,7141788,7141980,7139704,7141792,7143237,7142238,
max: 14, chi_sq: 12.860781, exp: 7142857.142857, jmax: 6
max =  15: 6664666,6667752,6665883,6667425,6671139,6667166,6666848,6669380,6662383,6664535,6663906,6663461,6665028,6668392,6672036,
max: 15, chi_sq: 16.394865, exp: 6666666.666667, jmax: 7
```

File: include/internal/common.h
Comment (mspncp) [inline include/internal/common.h:41]: Why is this global change hidden in here? It should be raised as a separate pull request imo.

**Review by mspncp — DISMISSED** [review]

**Comment by mspncp** [comment]
nit: unbiassed -> unbiased

File: include/internal/common.h
Comment (paulidale) [reply include/internal/common.h]: #22516

**Comment by paulidale** [comment]
Addressed comments...

**Comment by paulidale** [comment]
Ping @openssl/committers for reviews.

The algorithm isn't as daunting as it first appears.

File: include/crypto/rand.h
Comment (tom-cosgrove-arm) [inline include/crypto/rand.h:145]: Where would documentation be for this? I presume we are after a number n such that lower <= n < upper, but that should be clearly documented somewhere.

File: crypto/rand/rand_uniform.c
Comment (tom-cosgrove-arm) [inline crypto/rand/rand_uniform.c:70]: Given that `ossl_rand_uniform_uint32()` is going to assert if `upper - lower` (on line 74) == 0 should we not require that `lower` is strictly `> upper` here?

File: crypto/rand/rand_uniform.c
Comment (tom-cosgrove-arm) [inline crypto/rand/rand_uniform.c:23]: Why not just call this `max_followup_iterations` and avoid the comment?

Comment (paulidale) [reply crypto/rand/rand_uniform.c:23]: Good suggestion.

File: crypto/rand/rand_uniform.c
Comment (paulidale) [reply crypto/rand/rand_uniform.c:70]: It could be done that way but I figured failing earlier was reasonable.
It's not like there is a cost difference between `>=` and `>` on mot processors.

Also, it didn't seem worthwhile introducing an extra function which both others used.

File: include/crypto/rand.h
Comment (paulidale) [reply include/crypto/rand.h:145]: Done in the header file.

File: crypto/rand/rand_uniform.c
Comment (tom-cosgrove-arm) [reply crypto/rand/rand_uniform.c:71]: Err, my point is to change this to

```
    if (!ossl_assert(lower > upper)) {
```

in which case we will fail earlier (at this point) if lower == upper

Comment (paulidale) [reply crypto/rand/rand_uniform.c]: Yeah, it should have been that way.

A test that calls both and confirms the results are in range would be possible and worthwhile.  I'll add one.
Testing the uniformity isn't practical.

Comment (paulidale) [reply crypto/rand/rand_uniform.c]: Test added.

File: test/rand_test.c
Comment (tom-cosgrove-arm) [inline test/rand_test.c:71]: Should there be negative tests? i.e. that we get an error if we call either function with invalid parameters?

File: crypto/rand/rand_uniform.c
Comment (tom-cosgrove-arm) [reply crypto/rand/rand_uniform.c]: > A test that calls both and confirms the results are in range would be possible and worthwhile. I'll add one.

Thanks

> Testing the uniformity isn't practical.

Agreed

**Comment by tom-cosgrove-arm** [comment]
@mspncp Are you happy that your requested changes have been made?

**Comment by mspncp** [comment]
Yes I am. You may dismiss my previous change request if you like. (I didn't find a way how to do it from the mobile app.)

File: test/rand_test.c
Comment (paulidale) [reply test/rand_test.c:71]: They'll abort due to the assert.  So no.

**Review by tom-cosgrove-arm — APPROVED** [review]
LGTM

**Comment by paulidale** [comment]
Ping @mspncp or @openssl/otc for review please.

**Comment by paulidale** [comment]
Adding 3.2 because the HPKE change is worthwhile there and it depends on this.

**Comment by tom-cosgrove-arm** [comment]
> Adding 3.2 because the HPKE change is worthwhile there and it depends on this.

Can I get a clarification here? [This comment from @mattcaswell](https://github.com/openssl/openssl/pull/22450#issuecomment-1781314451) says that beta releases are usually treated as stable releases - would new features go there?

File: test/rand_test.c
Comment (tom-cosgrove-arm) [reply test/rand_test.c:71]: It would be nice if there was a way in the test framework to have "expect abort", so these sort of things could be tested, but that's not for this PR

**Comment by paulidale** [comment]
The HPKE fix has been deemed a bug fix which means it can be back ported.  It requires this, so this will be too.
Beta (hopefully) freezes the public API.  Since this is entirely internal it not part of our API guarantee.

**Review by t8m — APPROVED** [review]

File: crypto/rand/rand_uniform.c
Comment (mspncp) [inline crypto/rand/rand_uniform.c:17]: I'd prefer a complete description of the algorithm here in the source code instead just a link to a pull request in another library. I doesn't have to be as detailed as the original comments, but contain the main essence.

File: crypto/rand/rand_uniform.c
Comment (mspncp) [inline crypto/rand/rand_uniform.c:44]: Just a wild thought, in view of the timing comments https://github.com/apple/swift/pull/39143#issuecomment-912185936: 

Since the block size of our DRBG is 128 bits, one could make a single generate call for 128 bits, which would yield input for an unconditional calculation using a 32bit integer and 96bit fractional part (up to 3 followup iterations in your wording). Wouldn't that be good enough in practice to avoid further looping?

**Review by mspncp — CHANGES_REQUESTED** [review]
The algorithm looks indeed very promising and interesting. But since the random algorithm is intended to be used for cryptographic purposes, where good randomness is crucial, we shouldn't rush it in light-heartedly.

In particular, I'd like to ask some nasty questions:


### How trustworthy is the algorithm?

The pull request https://github.com/apple/swift/pull/39143 was raised in September 2021. It contains a very detailed and convincing explanation, but contrary to [@lemire's algorithm](https://arxiv.org/abs/1805.10941) which it attempts to replace, there is no official publication of the algorithm in a mathematical journal. The pull request has received a lot of positive feedback, even a comment by @lemire, but after more than two years it is still in draft mode and unmerged. How well reviewed is the algorithm and how trustworthy is it?

### What about timing issues?

As mentioned in https://github.com/apple/swift/pull/39143#issuecomment-912185936, the algorithm can leak timing information. Can this potentially cause problems for the intended use-case, HPKE?


### What about the CLA?

I have a question concerning the CLA for which I would like to get an official statement from the @openssl/omc:

Our [CLA policy](https://www.openssl.org/policies/cla.html) explicitly states that

> Every non-trivial contribution needs to be covered by a signed Contributor License Agreement (CLA) from all original authors. 

The implementation was taken from the Swift project and the original author @stephentyrone has not signed a CLA with the OpenSSL project. The Swift project however happens to be covered by an Apache License v2 as well, so one might infer from the Apache License text and CLA text that by signing a CLA with the Swift project the original author has indirectly granted the Copyright License to the OpenSSL project and no CLA is required.

Is that the case, and if yes, where is is it documented in our policies? Does this reasoning only apply to Apache licenses? I know that similar arguments have been rejected in the past for contributions which were declared as public domain by their author in order to avoid signing a CLA.

As indication that the CLA situation is not obvious to me, I've added the CLA label.

### Should this go into 3.2 as a last-minute change?

I'm not sure. Given the fact that we are switching to time-based releases, it might not hurt to wait until 3.3.

**Comment by sftcd** [comment]
Hiya,

A good set of questions, worth asking...

On 31/10/2023 00:52, Matthias St. Pierre wrote:
> ### What about timing issues? As mentioned in 
> https://github.com/apple/swift/pull/39143#issuecomment-912185936, the 
> algorithm can leak timing information. Can this potentially cause 
> problems for the intended use-case, HPKE?

I can try answer that part. As of now, the only coded-up use of this
is for GREASEing HPKE which is so far only for ECH, and only for use on
TLS clients before sending a ClientHello, so I don't think there's a
timing concern there for any realistic client using OpenSSL.

That said, with ECH, I have a TODO which is to add some GREASEing of
``retry_config`` values, sent in response to a TLS ClientHello as an
encrypted extension. In theory, that could emit a timing signal, but
my plan is to do that GREASEing (and hence call this RNG) at key load
time and not at response generation time. (That code is not written
yet, so that could change I guess, but I'm happy to say I won't do
that particular bad thing when making a PR for ECH:-)

I think QUIC+ECH would/should be the same as TLS+ECH in terms of the
above, but there's no actual QUIC+ECH implementation so far.

The other HPKE-using specification that might warrant consideration is
MLS. I've not done any analysis of how timing issues might arise there
but if this were also only used for GREASEing then the situation should
be similar.

For this PR, I'd say adding a note to the documentation that says "do
consider timing issues" might be sufficient at this point, given that
the function is so far only being used to select GREASE'y values and
not values that are later used as keys or similar. Having that
documentation describe whatever timing issues may exist would be nice.

Lastly, for HPKE, the alternative to this is to emit packets with a
minor bias in GREASE'y codepoints (the status-quo ante), or to do
rejection sampling, and I'd guess either would emit more of a signal
(via timing or codepoints) to network observers than this algorithm.

If it helps, I'm happy to raise the GREASE'y codepoint timing issue
with relevant IETF lists, but I suspect the answer we get will be that
there are so many other timing and/or traffic analysis signals in
protocols using HPKE that this is not currently a live concern And I
think that'd likely be a correct conclusion.

Cheers,
S.

PS: In case it helps, the GREASE'y use of this for ECH is not to
pick values like fake keys or ciphertext values, but only to pick
fake algorithm/suite IDs. Subsequent code then generates "real"
fake values for keys/ciphertext based on those not-really-in-use,
but usable, algorithm/suite IDs.

**Comment by paulidale** [comment]
IMO a CLA is not required.  I wrote **all** of this code myself.  I did not base it on the referenced code.

More specifically, the reference code generates at most two 64 bit integers and assumes the final error is less than 2^-64.
Here, I keep generating 32 bit integers until there is a final definite resolution (at least until the loop terminates via the counter).  I figured things out from the description and extended it to multiple limbs.

Thus, I do not believe the CLA: require tag is appropriate.

File: crypto/rand/rand_uniform.c
Comment (paulidale) [reply crypto/rand/rand_uniform.c:44]: If we could guarantee a 128 bit integer type we could do this.
We need an integer that can be multiplied by the range and stay within limits.
IMO: 32 bits is the best we can hope for.

Although we require 64 bit integers we do not mandate 128 bit multiplication results.  Most modern compilers support such and cannot rely on it being available.  IMO, the time critical part of this code will be the random generator not the reduction to range.

File: crypto/rand/rand_uniform.c
Comment (paulidale) [reply crypto/rand/rand_uniform.c:17]: Okay, can do but not today.

**Comment by paulidale** [comment]
In response to the other questions:

> How trustworthy is the algorithm?

Very.  It is generating a uniform [0, 1) value which is multiplied by the range.
If there isn't a definitive outcome based on the fractional part of the range, it generates more bits until there is.
So long as the RNG is unbiassed, the result will be.

I have no concerns here.


> What about timing issues?

Irrelevant IMO.  Multiple steps generating a random number reveals nothing about the final result.  All an attacker could learn was that the generated number's fractional part was just under one half.  For r ∈ [0, n), you know that frac(r) was either just below or equal to ½.  Assuming that rounding down is the final outcome (which it almost always will be), this gives you no information about ⌊r⌋, just that it cannot be _n_.


> Should this go into 3.2 as a last-minute change?

This should go in if the associated HPKE PR goes in.  If that doesn't go in, this shouldn't.
It's not a public API so it is not a breaking change.

I believe there is an OTC hold in place, so we'll discuss later (assuming I'm awake).

**Comment by t8m** [comment]
This was written from scratch. CLA hold does not apply.

**Comment by t-j-h** [comment]
**OTC: vote to include PR subject to normal review process**

**Comment by mspncp** [comment]
> This was written from scratch. CLA hold does not apply.

I didn't doubt the fact that it was written from scratch. But except for minor adjustments, it is the same algorithm. As a mathematician, I find it remarkable and non-intuitive that it's only the *written program* and not the underlying *algorithm* which is considered an intellectual property. Fortunately, I'm not a lawyer, and if they are happy with this interpretation, I'll not dive down further into this rabbit hole.

File: crypto/rand/rand_uniform.c
Comment (mspncp) [reply crypto/rand/rand_uniform.c:44]: I wasn't thinking about using larger integers, only about joining the first four `RAND_bytes()` calls into a single call . But since the 128bits are cached internally and unused bits are consumed later iirc, this optimization might not bring such a large improvement as I initially thought.

**Comment by paulidale** [comment]
The code here is different to the one referenced.  The referenced one only generates a second 64 bit value and accepts a 1 in 2^64 chance of bias.  This PR, instead, generates 32 bit values and continues to do so until a definitive result is obtained -- this is not covered by the referenced code.

Algorithms cannot be copyrighted.  They are patentable (unfortunately).

File: crypto/rand/rand_uniform.c
Comment (paulidale) [reply crypto/rand/rand_uniform.c:44]: I misunderstood.

The DRBG cache their output, so I think you're correct in that that change wouldn't help much.

We don't actually want lots of bits.  The probability of needing a third block of 32 bits is under 2¯³³.  The probability of needing the second block is at most ½ and usually far less.

File: crypto/rand/rand_uniform.c
Comment (paulidale) [reply crypto/rand/rand_uniform.c]: Outline of algorithm added.

File: crypto/rand/rand_uniform.c
Comment (mspncp) [inline crypto/rand/rand_uniform.c:48]: but -> by

File: crypto/rand/rand_uniform.c
Comment (mspncp) [inline crypto/rand/rand_uniform.c:69]: repreat -> repeat

**Review by t8m — APPROVED** [review]
Still ok, no need to restart the 24h timer

**Comment by t8m** [comment]
@tom-cosgrove-arm please reconfirm

**Review by mspncp — APPROVED** [review]
Nice work done on the comments. Just two typos.

File: crypto/rand/rand_uniform.c
Comment (paulidale) [reply crypto/rand/rand_uniform.c]: Hmmmm....I ran this through a spell checker....

All fixed now.

**Review by t8m — APPROVED** [review]
Still OK, no need to restart the timer

**Comment by tom-cosgrove-arm** [comment]
Still okay

**Comment by t8m** [comment]
Merged to the master and 3.2 branches. Thank you.

**Pull request closed without merging**
